Add JsDoc for Class Contact

lijiarui opened the pull request:
#73 
#252 
use JsDoc to generate contact doc here.
Doc: [Contact](https://github.com/lijiarui/wechaty/blob/testDoc/docs/index.md#genderundefined-2)

zixia commented on src/contact.ts:
```diff
@@ -55,9 +64,8 @@ export type ContactQueryFilter = {
- * blabla...
- * **IMPORTANT**
  *
+ * `Contact` is `Sayable`
```
Please add example to each method.

like:

```ts
/**
 * @example
 * ```ts
 *   code here
 * ```
 */

lijiarui replied:
I don't think we should add each method, some method like `Contact.name()`, the code already can explain itself as `Gender` you said.

I have added method `Contact.find()` in this commit.

But your suggestion make sense, except method `find`, I think I should add the following method:
- avatar()
- say()

What other functions do you think I should add too?

zixia replied:
I do not agree. Document is not the same as Code.

Examples can help users to understand easier how to use it rather than the API reference.

Sometimes you think it's obviously but for others are not, that's because you are too familiar with Wechaty.

I think to add examples for all API functions is necessary, but I'll leave this decision to you.

Please add examples to the APIs which you think is needed, leave it without example as you like.

lijiarui replied:
You are right, maybe I'm too familiar with Wechaty, so I add examples for all API functions.

zixia commented on src/contact.ts:
```diff
@@ -107,20 +115,76 @@ export class Contact implements Sayable {
   public name()     { return UtilLib.plainText(this.obj && this.obj.name || '') }
+
+  /**
+   * Check if contact is strange
```
is stranger.

zixia commented on src/contact.ts:
```diff
@@ -223,6 +302,13 @@ export class Contact implements Sayable {
+   * @static
+   * @param {ContactQueryFilter} [queryArg]
+   * @returns {Promise<Contact[]>}
+   * @description If use Contact.findAll() get the contact list of the bot.
```
Put @description at the beginning of the comment block.

Because the comments at the beginning of the comment block is default to be @description.

zixia commented on src/contact.ts:
```diff
@@ -292,16 +378,31 @@ export class Contact implements Sayable {
+   *
+   * @returns {(string | null)}
+   *
+   * @memberOf Contact
```
`@memberOf` is not needed in ES6/Typescript.

Because the JsDoc is already known it is member of the class.

lijiarui replied:
So your suggestion is to delete all `@memberOf` ?

zixia replied:
Yes. 

Unless you have a reason for keeping it.

zixia commented on src/contact.ts:
````diff
@@ -349,8 +450,17 @@ export class Contact implements Sayable {
+   * ```
+   * const contactFindByName = await Contact.find({ name:"ContactName"} )
+   * const contactFindByAlias = await Contact.find({ alias:"ContactAlias"} )
+   * ```
````
specify the type of the example.

here should be:

\`\`\`ts

lijiarui commented on src/contact.ts:
```diff
@@ -40,9 +40,18 @@ export type ContactRawObj = {
+/**
+ * Enum for Gender values.
+ * @readonly
+ * @export
```
You are right, keep it simple, only list the useful thing.

lijiarui commented on docs/index.md:
```diff
@@ -205,44 +259,170 @@ Add/Del/Topic: https://github.com/wechaty/wechaty/issues/32
+**Export**:   
 <a name="config_1"></a>
 
 ## config_1
```
After I modified as you requested, it has changed automatically, maybe it is correct now.

zixia commented on docs/index.md:
```diff
@@ -25,9 +25,8 @@ Wechaty.instance() // Singleton
 <p>Add/Del/Topic: <a href="https://github.com/wechaty/wechaty/issues/32">https://github.com/wechaty/wechaty/issues/32</a></p>
 </dd>
 <dt><a href="#Contact">Contact</a></dt>
-<dd><p>Class Contact
```
Do not commit the new docs/index.md in this PR.

Because this is not necessary.

Please only commit the code file.

lijiarui replied:
I won't commit docs/index.md in this PR any more.

zixia commented on src/contact.ts:
```diff
@@ -107,20 +110,76 @@ export class Contact implements Sayable {
+   *
+   * @returns {string | ''}
+   *
+   * @memberOf Contact
```
?

lijiarui replied:
I didn't commit yet after your comment...

zixia commented on src/contact.ts:
```diff
@@ -107,20 +110,76 @@ export class Contact implements Sayable {
+   *
+   * @returns {string | ''}
+   *
+   * @memberOf Contact
```
@memberOf is not necessary as we discussed.

lijiarui commented:
@zixia 
done all of your request.

- @memberOf  removed  all `@memberOf`
- example:
   You are right, maybe I'm too familiar with Wechaty, so I add examples for all API functions.

zixia commented:
No, you are not done.

Please mark all my reviews as `resolved` after you confirmed it is resolved.

I noticed that there has some reviews you did not reply/react.

lijiarui commented:
How about this time?

zixia commented:
I don't know.

Please tell me you are done after you are done.

lijiarui commented:
done

lijiarui commented:
I'm sorry....

I thought I used wrong git command...
I should use `git rm –cached` instead of `git rm` to delete index instead of delete both index and the file.

But I thoug your bot can generate the docs automatically.

zixia commented:
No, you are not done.

1. you leave one un-resolved review.
2. you should not commit the file docs/index.md. this time you still commit it, with worse behaviour: deleted it.

lijiarui commented:
done

zixia commented:
Not yet.

zixia commented on src/contact.ts:
```diff
@@ -107,20 +110,103 @@ export class Contact implements Sayable {
+  /**
+   * Get the weixin number from a contact
+   * Sometimes cannot get weixin number due to weixin security mechanism, not recommend.
+   * @returns {string | ''}
```
@returns {string}

or {string|null} is better, for null is standing for we can not get the wexin id.

zixia commented on src/contact.ts:
````diff
@@ -107,20 +110,103 @@ export class Contact implements Sayable {
+   * const weixin = contact.weixin()
+   * ```
+   */
   public weixin()   { return this.obj && this.obj.weixin || '' }
````
public weixin(): string
or 
public wexin(): string | null

zixia commented on src/contact.ts:
```diff
@@ -107,20 +110,103 @@ export class Contact implements Sayable {
+  /**
+   * Get the name from a contact
+   *
+   * @returns {string | ''}
```
@returns {string}

zixia commented on src/contact.ts:
````diff
@@ -107,20 +110,103 @@ export class Contact implements Sayable {
+   * const isStranger = contact.stranger()
+   * ```
+   */
   public stranger() { return this.obj && this.obj.stranger }
````
public stranger(): boolean

and make sure the function is return a type of `boolean`

lijiarui commented:
I'm not sure which haven't done....

> 1.you leave one un-resolved review.
> 2.you should not commit the file docs/index.md. this time you still commit it, with worse behaviour: deleted it. I need your PR only include one file modification instead of two.

1. I have replyed your `example review`
2. I have recovered file I deleted

So could you tell me more? 
Thanks.

zixia commented on src/contact.ts:
````diff
@@ -107,20 +110,103 @@ export class Contact implements Sayable {
+   * const isStar = contact.star()
+   * ```
+   */
   public star()     { return this.obj && this.obj.star }
````
public star(): boolean

and make sure the function is return a type of boolean

zixia commented on src/contact.ts:
````diff
@@ -107,20 +110,103 @@ export class Contact implements Sayable {
+   * const gender = contact.gender()
+   * ```
    */
   public gender()   { return this.obj ? this.obj.sex : Gender.Unknown }
````
public gender(): Gender

zixia commented:
Last time you leave this review **unresolved**:

![image](https://cloud.githubusercontent.com/assets/1361891/23840980/da58580e-07e4-11e7-9c26-4441156779d6.png)

lijiarui commented:
So how should I resolve this review now ....

lijiarui commented:
done ` fulfill all the return type for functions`...
But I'm not sure whether I have resolved the example review... 
Could you tell me?
Thanks

zixia commented:
You still leave one review _unresolved_.

But it's ok, I'll merge this PR now.

zixia commented:
BTW: there's a bug that causing the unit test failed, it seems related to the PR from you recently. 

Please have a look when you have time. Thanks.

lijiarui commented:
Could you tell me how to resolve the review? 
Thanks

zixia commented:
I do not know because I can not see your account.
